### PR TITLE
initial useKeypress

### PR DESCRIPTION
**Pull request by jedahan**
first draft of #138 

**todo**
- [x] fix typescript definition (never written typescript before)
- [x] run npm test
~- [ ] write documentation~
- [x] write example
- [x] write some tests

### CODE REVIEW

**Comment by sindresorhus** [comment]
This needs to handle the case if raw mode is already enabled when this hook is used, it should not be disabled when this hook ends.

**Comment by vadimdemedes** [comment]
That's actually going to be handled by `setRawMode()` itself, so hook doesn't need to re-implement that logic :)

File: src/hooks/useKeypress.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  const { stdin, setRawMode } = useContext(StdinContext)
+
+  useEffect(() => {
+    setRawMode(true)
+    stdin.on('keypress', keypressHandler)
+    return () => {
+      stdin.off('keypress', keypressHandler)
+      setRawMode(false)
+    }
+  }, [stdin, setRawMode])
```
Comment (SimenB) [inline src/hooks/useKeypress.js]: need to add `keypressHandler` as well to the dependency array

File: index.d.ts [hunk 2]
```diff
@@ -44,6 +44,8 @@ export type Instance = {
 
 export type Unmount = () => void;
 
+export function useKeypress<T extends (str?: string, key?: any) => null>(keyPresshandler: T): null;
```
Comment (SimenB) [inline index.d.ts]: type for `key`: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/de6730f4463cba69904698035fafd906a72b9664/types/node/readline.d.ts#L5-L11

You can import it by `import {Key} from 'readline';`

Should also add a dependency on `@types/node` - I forgot in #1

File: src/hooks/useKeypress.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { useEffect, useContext } from 'react'
+import StdinContext from '../components/StdinContext'
+
+// TODO: memo? useRef? Will this re-init on render?
+// Should this re-initialize only if the keypressHandler changes?
```
Comment (SimenB) [inline src/hooks/useKeypress.js]: shouldn't be necessary for you to think about, just putt everything you pass into the closure in the dep array

File: index.d.ts (hunk 2 above)
Comment (SimenB) [reply index.d.ts]: Maybe not the dependency and just hope consumers add it if they need it? IIRC the error from `tsc` is ok (it specifically calls out installing `@types/node`)

The global nature of TS's types seems to cause issues https://github.com/facebook/jest/issues/8092

File: test/keypress.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+				t.deepEqual(hits, input);
+				t.end();
+			}
+		});
+		return null;
+	};
+
+	render(<KeypressTest/>);
+	for (const letter of input.split('')) {
+		process.stdin.push(Buffer.from(letter));
```
Comment (sindresorhus) [inline test/keypress.js]: It's more correct to use `process.stdin.write()`. Then you don't need to wrap it in a Buffer either.

File: index.d.ts [hunk 5]
```diff
@@ -141,11 +144,11 @@ export interface BoxProps {
...
-		| "flex-start"
-		| "center"
-		| "flex-end"
-		| "space-between"
-		| "space-around";
+	| "flex-start"
+	| "center"
+	| "flex-end"
+	| "space-between"
+	| "space-around";
```
Comment (sindresorhus) [inline index.d.ts]: Don't do unrelated changes.

File: index.d.ts [hunk 6]
```diff
@@ -44,6 +45,8 @@ export type Instance = {
 
 export type Unmount = () => void;
 
+export function useKeypress<T extends (str?: string, key?: Key) => null>(keyPresshandler: T): null;
```
Comment (sindresorhus) [inline index.d.ts]: No need to use generics here.

Comment (sindresorhus) [reply index.d.ts]: It returns `void`, not `null`.

**Comment by jedahan** [comment]
So I got stuck on my test - it would be cool if ava could allow you to get a handle to stdin/stdout or the subprocess when running tests so we can simulate keypresses etc.

File: index.d.ts (hunk 2 above)
Comment (jedahan) [reply index.d.ts]: Is the current version okay regarding types? Is there a particular class/book/tutorial you recommend on learning the ins and outs of practical typescript?

**Comment by vadimdemedes** [comment]
See https://yarnpkg.com/en/package/node-pty on how you can write to stdin of the process created by `node-pty` (which is used in `test/exit.js`) ;)

**Comment by vadimdemedes** [comment]
Hey @jedahan, do you have time to finish this PR? Totally ok if not ;)

**Comment by jedahan** [comment]
Unfortunately I cannot get `master` to npm install on my laptop now. Which is weird because a fresh checkout of node-pty 0.8.1 does npm install correctly, but thats where its erroring

**Comment by vadimdemedes** [comment]
Try using Node.js v10.x.

**Comment by sindresorhus** [comment]
@jedahan Did you manage to get it to npm install?

**Comment by jedahan** [comment]
`npx -p node@10 npm install && npx -p node@10 npm test` passes

**Comment by jedahan** [comment]
Working on fixing the test now

File: examples/usekeypress/usekeypress.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+			setY(Math.max(1, y - 1));
+		}
+
+		if (str === 'q') {
+			exit();
+		}
+	});
+
+	// eslint-disable-next-line react/no-array-index-key
+	const ypad = new Array(y).fill().map((_, i) => <span key={i}> <br/></span>);
```
Comment (sindresorhus) [inline examples/usekeypress/usekeypress.js]:
```suggestion
	const yPad = new Array(y).fill().map((_, i) => <span key={i}> <br/></span>);
```

File: examples/usekeypress/usekeypress.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+	const paddedface = '^_^'.padStart(x+3);
+
+	return (
+		<div>
```
Comment (sindresorhus) [inline examples/usekeypress/usekeypress.js]: This should use `<Box>`

File: examples/usekeypress/usekeypress.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+const React = require('react');
+const {render, useKeypress} = require('../..');
+
+const {exit} = process;
+
+const Robot = () => {
+	const [x, setX] = useState(1);
+	const [y, setY] = useState(1);
+
+	useKeypress(str => {
```
Comment (sindresorhus) [inline examples/usekeypress/usekeypress.js]:
```suggestion
	useKeypress(key => {
```

File: index.d.ts [hunk 10]
```diff
@@ -1,4 +1,5 @@
 import * as React from "react";
+import { Key } from 'readline';
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
import {Key} from 'readline';
```

File: index.d.ts [hunk 11]
```diff
@@ -44,6 +45,14 @@ export type Instance = {
 
 export type Unmount = () => void;
 
+/**
+ * Hook that calls keyPressHandler callback with whatever key has been pressed.
+ * See key to handle modifiers correctly.
+ */
+export function useKeypress(
+	keyPressHandler: (str?: string, key?: Key) => void
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
	keyPressHandler: (string?: string, key?: Key) => void
```

File: index.d.ts [hunk 12]
```diff
@@ -44,6 +45,14 @@ export type Instance = {
 
 export type Unmount = () => void;
 
+/**
+ * Hook that calls keyPressHandler callback with whatever key has been pressed.
+ * See key to handle modifiers correctly.
```
Comment (sindresorhus) [inline index.d.ts]: See what? Do you mean `Key`?

File: test/fixtures/handles-keypresses.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+		if (input.size === 0) {
+			app.unmount();
+		}
+	});
+	return null;
+};
+
+const app = render(<KeypressTest/>);
+
+app.waitUntilExit().then(() => console.log('exited'));
```
Comment (sindresorhus) [inline test/fixtures/handles-keypresses.js]:
```suggestion
(async () => {
	await app.waitUntilExit();
	console.log('Exited');
})();
```

File: test/fixtures/handles-keypresses.js [hunk 14]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+const React = require('react');
+const {render, useKeypress} = require('../..');
+
+const input = new Set('abcdefghijklmnopqrstuvwxyz'.split(''));
+
+const KeypressTest = () => {
+	useKeypress(str => {
```
Comment (sindresorhus) [inline test/fixtures/handles-keypresses.js]:
```suggestion
	useKeypress(string => {
```

File: examples/usekeypress/usekeypress.js (hunk 9 above)
Comment (jedahan) [reply examples/usekeypress/usekeypress.js]: so useKeypress calls the callback with the string representation of the key, plus the Key object `(str, key)`. In this case, we just care about the string representation. Its might be a bit confusing and we should just choose the full Key object representation, allowing people to destructure with `({ str })` perhaps. What are your thoughts

Comment (vadimdemedes) [reply examples/usekeypress/usekeypress.js]: What are the use cases for wanting the full object, not just the string representation?

Comment (SimenB) [reply examples/usekeypress/usekeypress.js]: if `ctrl` is held down etc I'd guess?

Comment (vadimdemedes) [reply examples/usekeypress/usekeypress.js]: It could be passed as a second argument to the callback in that case:

```js
useKeypress((key, { ctrl }) => ...)
```

Comment (SimenB) [reply examples/usekeypress/usekeypress.js]: That's what happens with the current implementation

Comment (SimenB) [reply examples/usekeypress/usekeypress.js]: key object is probably more useful, btw.

```js
const readline = require('readline');
readline.emitKeypressEvents(process.stdin);
process.stdin.setRawMode(true);
process.stdin.on('keypress', console.log);
```

```sh-session
$ node script.js
# hit s
s { sequence: 's',
  name: 's',
  ctrl: false,
  meta: false,
  shift: false }
# hit shift + s
S { sequence: 'S', name: 's', ctrl: false, meta: false, shift: true }
# hit ctrl + s
 { sequence: '\u0013',
  name: 's',
  ctrl: true,
  meta: false,
  shift: false }
```


So I think just a single `Key` object might be better?

Comment (vadimdemedes) [reply examples/usekeypress/usekeypress.js]: I'd keep first argument as a key string for convenience to avoid destructuring when you only care about what key was pressed and keep `Key` object as a 2nd argument with `ctrl`, `shift`, etc props.

Comment (jedahan) [reply examples/usekeypress/usekeypress.js]: We could experiment with alternative, simpler apis and how they feel with higher level hooks maybe like ‘useKey = cb => useKeypress((_, key) => cb(key))’

Other ways to interpret the key object into a string might be a good middle ground as well.

‘useKeystring = cb => useKeypress((_, key) => cb(keyToString(key)))’

Where keyToString converts the key object to strings like ‘meta+shift+s’ or ‘meta+S’.

**Comment by jedahan** [comment]
I'm rebasing

**Comment by jedahan** [comment]
I think I've made all the requested changes

**Comment by vadimdemedes** [comment]
Hey @jedahan, I've tried running `node examples/usekeypress` and process exits right away, after rendering the output. It doesn't wait for user input.

**Comment by vadimdemedes** [comment]
I digged deeper and was able to fix it by replacing `useEffect` with `useLayoutEffect`. The problem with `useEffect` is that it fires the handler after rendering is finished, so `setRawMode` isn't called fast enough. Process sees that there are no more tasks left and exits.

I also decided to switch from using `keypress` to `data` event, because when user pastes text into terminal, multiple `keypress` events are emitted for each character. This is not as convenient as using `data` event, which contains the whole string in one event.

I'll open another PR, which includes your changes and fixes these things. I will make sure to mention you, of course.

**Comment by vadimdemedes** [comment]
Opened https://github.com/vadimdemedes/ink/pull/227 to take this PR further. I would've asked @jedahan, but I want to release it asap.
